text/0063-coprocessor-plugin.md

```diff
@@ -0,0 +1,323 @@
+    async fn batch_delete(&self, keys: Vec<Key>) -> PluginResult<()>;
+
+    /// Same as [`RawStorage::delete()`], but deletes multiple key-values pairs at once
+    /// given a `key_range`. All records with keys in `[key_range.start, key_range.end)`
+    /// will be deleted. The upper bound of the `key_range` is exclusive.
+    async fn delete_range(&self, key_range: Range<Key>) -> PluginResult<()>;
+}
+```
+
+Build the plugin **using the same `rustc` as TiKV are built** and then you will find the dylib in `target/debug` or `target/release`. Next, copy the `.so` (Linux) or `.dylib` (MacOS) file to the `coprocessors/` folder next to `tikv-server`, and then TiKV will load it automatically on startup or even while running. The `coprocessors/` folder is configurable by TiKV config file:
+
+```toml
+[coprocessor-v2]
+## Path to the directory where compiled coprocessor plugins are located.
+## Plugins in this directory will be automatically loaded by TiKV.
+coprocessor-plugin-directory = "./coprocessors"
+```
+
+### Multi-plugin
+
+Multiple plugin can be load at the same time as long as the plugins have different names or versions.
+
+### Auto loading
+
+TiKV will watch the plugin directory and load new plugin when new plugin file is created. New plugin should have new file name and should not have the same plugin name and version number as the already loaded plugin.
+
+## Reference-level design
+
+### Plugin runtime
+
+The plugin runtime is a new component settling in `tikv::server::service::kv::Service`. It loads the dylib plugin, dispatches coprocessor request to the plugin, and proxy the API calls from plugin to the [`Storage`](https://tikv.github.io/doc/tikv/storage/struct.Storage.html).
+
+The path of the plugin should be specified in the config file and be loaded at TiKV startup.
+
+### Auto Loading
+
+It should be convenient to be able to hot reload the plugin without restarting TiKV. But due to the dylib constraints described in [`rust-hot-reload`](https://github.com/irh/rust-hot-reloading#avoiding-thread-local-storage):
+
+> Any use of TLS will prevent hot-reloading from working, even with the unique copy trick.
+
+The hot-reload plugin should have new filename to avoid conflict with the existing plugins, which means the plugins will never be unload by TiKV, but you can register a new plugin with new version number or new name.
+
+### Protobuf design
+
+```proto
+message RawCoprocessorRequest {
+    kvrpcpb.Context context = 1;
+
+    string copr_name = 2;
+    // Coprorcessor version constraint following SEMVER definition.
+    string copr_version_req = 3;
+
+    repeated KeyRange ranges = 4;
+    bytes data = 5;
+}
+
+message RawCoprocessorResponse {
+    errorpb.Error region_error = 1;
+    // Error message for cases like if no coprocessor with a matching name is found
+    // or on a version mismatch between plugin_api and the coprocessor.
+    string error = 2;
+    bytes data = 3;
+}
+```
+
+### Plugin API design
```

Comment:
except for the API that a plugin needs to implement, on the user side, how to:

- manage plugins: observe existing plugins, register new plugins, deregistration existing plugins, plugin privileges (plugins can be added by different people for different purposes)
- utilize plugins in the TiKV client API: how to scan keys that satisfy a user-defined filer based on user-defined data types and key-value codecs?